diff --git a/setup.sh b/setup.sh
index f1f85fd..8d7c843 100755
--- a/setup.sh
+++ b/setup.sh
@@ -1,13 +1,30 @@
 #!/usr/bin/env bash
 
-git submodule update --init || exit 1
-cd XED-to-XML
-./mfile.py --opt=2 --no-encoder pymodule || exit 1
-cp xed.* ..
-cd ..
-git submodule deinit -f --all
-rm -rf .git/modules/*
+# This script does the same as the original setup.sh
+# but parallelizes the download with compilation
+
+set -e
+
+generate_xed() {
+  cd ./XED-to-XML
+  ./mfile.py --opt=2 --no-encoder pymodule
+  cp ./xed.* ..
+}
+
+download_instructions() {
+  wget https://www.uops.info/instructions.xml --quiet
+}
 
-wget https://www.uops.info/instructions.xml || exit 1
-./convertXML.py instructions.xml || exit 1
-rm instructions.xml
+git submodule update --init
+generate_xed &
+download_instructions &
+
+# wait for the download and xed file jobs to be done
+wait
+
+# convert
+./convertXML.py instructions.xml
+
+# cleanup
+git submodule deinit -f --all
+rm -rf .git/modules/* instructions.xml
